### PR TITLE
storage: Fix checkSequenceCache to propery set Error.Txn

### PR DESCRIPTION
**Pull request by kkaneda**
- Set `pErr.Txn` for `TransactionAbortedError`.
- Use `retryErr.Txn` to set `pErr.Txn` for `TransactionRetryError`. `retryErr.Txn` has advanced time while `txn` doesn't.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3938)

### CODE REVIEW

File: storage/replica_test.go [hunk 1]
```diff
@@ -2240,6 +2240,51 @@ func TestSequenceCachePoisonOnResolve(t *testing.T) {
...
+
+	txn := roachpb.Transaction{}
+	txn.ID = []byte("id")
+	txn.Sequence = 1
+	txn.Timestamp = roachpb.Timestamp{WallTime: 1}
+
+	// Populate the sequence cache with a higher sequence number
+	// to trigger TransactionRetryError.
+	key := roachpb.Key("k")
+	ts := txn.Timestamp.Next()
```
Comment (vivekmenezes) [inline storage/replica_test.go]: You might not need these two temp variables

Comment (vivekmenezes) [inline storage/replica_test.go]: nevermind

File: storage/replica_test.go [hunk 2]
```diff
@@ -2240,6 +2240,51 @@ func TestSequenceCachePoisonOnResolve(t *testing.T) {
...
+	}
+
+	// Poison the sequence cache to trigger TransactionAbortedError.
+	if err := tc.rng.sequence.Put(tc.engine, txn.ID, txn.Epoch, roachpb.SequencePoisonAbort, key, ts, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	pErr = tc.rng.checkSequenceCache(tc.engine, txn)
+	if err, ok := pErr.GoError().(*roachpb.TransactionAbortedError); ok {
+		if pErr.Txn == nil || !pErr.Txn.Equal(&err.Txn) {
```
Comment (tschottdorf) [inline storage/replica_test.go]: `Equal` compares only the ID. You want `reflect.DeepEqual`.

File: storage/replica_test.go [hunk 3]
```diff
@@ -2240,6 +2240,51 @@ func TestSequenceCachePoisonOnResolve(t *testing.T) {
...
+	// to trigger TransactionRetryError.
+	key := roachpb.Key("k")
+	ts := txn.Timestamp.Next()
+	if err := tc.rng.sequence.Put(tc.engine, txn.ID, txn.Epoch, txn.Sequence+1, key, ts, nil); err != nil {
+		t.Fatal(err)
+	}
+
+	pErr := tc.rng.checkSequenceCache(tc.engine, txn)
+	if err, ok := pErr.GoError().(*roachpb.TransactionRetryError); ok {
+		if pErr.Txn == nil || !pErr.Txn.Equal(&err.Txn) {
```
Comment (tschottdorf) [inline storage/replica_test.go]: ditto

**Comment by tschottdorf** [comment]
LGTM

File: storage/replica_test.go (hunk 2 above)
Comment (kkaneda) [inline storage/replica_test.go]: Ah, I see. Fixed.

File: storage/replica_test.go (hunk 3 above)
Comment (kkaneda) [inline storage/replica_test.go]: Fixed.

**Pull request merged**